Tools/ImageDiff/skia/PlatformImageSkia.cpp

```diff
@@ -0,0 +1,136 @@
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "PlatformImage.h"
+
+#include <skia/codec/SkPngDecoder.h>
+#include <skia/core/SkImage.h>
+#include <skia/core/SkPixmap.h>
+#include <skia/core/SkStream.h>
+#include <skia/encode/SkPngEncoder.h>
+#include <stdio.h>
+
+namespace ImageDiff {
+
+static std::unique_ptr<PlatformImage> createFromStream(std::unique_ptr<SkStream>&& stream)
+{
+    if (!stream)
+        return nullptr;
+
+    SkCodec::Result decodeResult;
+    auto codec = SkPngDecoder::Decode(std::move(stream), &decodeResult, nullptr);
+    if (decodeResult != SkCodec::Result::kSuccess) {
+        fprintf(stderr, "ImageDiff failed to decode PNG: %s\n", SkCodec::ResultToString(decodeResult));
+        return nullptr;
+    }
+
+    auto [image, result] = codec->getImage();
+    if (result != SkCodec::Result::kSuccess) {
+        fprintf(stderr, "ImageDiff failed to decode PNG: %s\n", SkCodec::ResultToString(result));
+        return nullptr;
+    }
+
+    return std::make_unique<PlatformImage>(std::move(image));
+}
+
+std::unique_ptr<PlatformImage> PlatformImage::createFromFile(const char* filename)
+{
+    return createFromStream(SkStream::MakeFromFile(filename));
+}
+
+std::unique_ptr<PlatformImage> PlatformImage::createFromStdin(size_t imageSize)
+{
+    SkDynamicMemoryWStream stream;
+    unsigned char buffer[2048];
+    size_t bytesRemaining = imageSize;
+    while (bytesRemaining > 0) {
+        size_t bytesToRead = std::min<size_t>(bytesRemaining, 2048);
+        size_t bytesRead = fread(buffer, 1, bytesToRead, stdin);
+        stream.write(buffer, bytesRead);
+        bytesRemaining -= static_cast<int>(bytesRead);
+    }
+    return createFromStream(stream.detachAsStream());
+}
+
+std::unique_ptr<PlatformImage> PlatformImage::createFromDiffData(void* data, size_t width, size_t height)
+{
+    auto imageInfo = SkImageInfo::Make(width, height, SkColorType::kGray_8_SkColorType, SkAlphaType::kOpaque_SkAlphaType);
+    SkPixmap pixmap(imageInfo, data, imageInfo.minRowBytes());
+    auto image = SkImages::RasterFromPixmap(pixmap, [](const void* pixels, void*) {
+        free(const_cast<void*>(pixels));
```

Comment:
How can we be sure free() is sufficient? could be fastalloc'ed data that's passed here, no?

Reply:
There's not fast malloc in ImageDiff